### PR TITLE
bash completion support

### PR DESCRIPTION
**Pull request by ecksun**
Simple bash completion support

### CODE REVIEW

File: notes.bash_completion [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+        [[ -d $item ]] && item="$item/"
+        local filename=${item#$notes_dir/}
+        COMPREPLY+=("${filename%.md}")
+    done
+}
+
+_notes()
+{
+    local cur
+    readonly commands="new find grep open"
```
Comment (pimterry) [inline notes.bash_completion]: When using this on my machine, I get lots of `bash: commands: readonly variable` when trying to use completion, which I assume is caused by this line. Completion still works, other than that spurious extra output. Using bash 4.4.5.

Any idea why? I'd like to fix this before merging.

File: README.md [hunk 2]
```diff
@@ -29,6 +29,18 @@ curl https://cdn.rawgit.com/pimterry/notes/v0.1.1/notes > /usr/local/bin/notes &
 
 By default your notes live in ~/notes, but you can change that to anywhere you like by setting the `$NOTES_DIRECTORY` environmental variable.
 
+### Bash completion
+
+If you want bash autocompletion copy the completion script into the bash completion directory. The bash completion directory is `/usr/share/bash-completion/completions/` on my system, you can you can get the bash completion directory by running the following command:
```
Comment (pimterry) [inline README.md]: Can you replace 'on my system' with some quick details of your system? E.g. `on a typical [ubuntu/debian/arch] install`.

File: notes.bash_completion [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+_notes_complete_notes() {
+    local configured_dir=${NOTES_DIRECTORY%/} # Remove trailing slashes
+    local notes_dir="${configured_dir:-$HOME/notes}"
+    local items=($(compgen -f "$notes_dir/$1"))
+    for item in "${items[@]}"; do
+        [[ -d $item ]] && item="$item/"
+        local filename=${item#$notes_dir/}
+        COMPREPLY+=("${filename%.md}")
+    done
```
Comment (pimterry) [inline notes.bash_completion:15]: This doesn't handle spaces correctly. If you have "my note.md" in your notes folder then `notes find <TAB><TAB>` will show completions for `my` and `note`.

Also, it includes matches for hidden files. That's a general `notes` issue I'd like to fix everywhere when I get a minute, but if it's easy it'd be nice to have it working properly here from the start.

**Review by pimterry — CHANGES_REQUESTED** [review]

**Comment by pimterry** [comment]
It's awesome to see people like you diving in on substantial improvements like this, this is great.

I've added a couple of comments. This also now needs rebasing, I think just because some of the README has changed in the meantime because of conflicts. Happy to merge once that's all done.

Thanks!

File: notes.bash_completion (hunk 3 above)
Comment (ecksun) [reply notes.bash_completion:15]: Hmm, I managed to fix it for spaces however I do not know how to deal with newlines in filenames still.

Should also not be that tricky to remove showing hidden files

File: notes.bash_completion (hunk 1 above)
Comment (ecksun) [reply notes.bash_completion]: This is very curious.

It seems to only become a problem the second time you press tab. This almost feels like a bug in bash to me, the variables should be scoped to the function yet they seem to survive multiple invocations, very strange.

I solved it by simply doing `local` instead, I will poke around a bit more to see if I can figure anything out.

**Comment by ecksun** [comment]
Now we have some tests, are handling whitespace notes and ignores hidden files.

Since I rebased the branch it seems I can no longer look at your review comments, nor mark tyour requested changes as completed (if that is something you can do with github).

**Comment by jacobmischka** [comment]
@pimterry you can rebase when accepting the PR, there's a dropdown. https://help.github.com/articles/about-pull-request-merges/

**Comment by pimterry** [comment]
@jacobmischka You can, but that's for a different use case, when you don't want a merge commit, and the PR doesn't currently have conflicts. You can't use that to avoid the rebasing required to fix conflicts in a PR (and I'd quite like to keep the merge commits, just for consistency).

@ecksun I think this looks good now. I need to do a proper test and review, and I don't have time today, but I should on Monday I think, so I'll take a proper look then, and merge it in. Great work, thanks!

**Review by pimterry — APPROVED** [review]

**Comment by pimterry** [comment]
Looks good, all works nicely for me locally, merged. Thanks @ecksun!

**Pull request merged**